sls-versions/src/main/java/com/palantir/sls/versions/SlsVersionMatcher.java

```diff
@@ -70,38 +71,32 @@ public static SlsVersionMatcher valueOf(String value) {
     /** The same as {@link #valueOf(String)}, but returns {@link Optional#empty} if the format is invalid. */
     public static Optional<SlsVersionMatcher> safeValueOf(String value) {
         checkNotNull(value, "value cannot be null");
+        return SlsVersionMatcherParser.safeValueOf(value);
+    }
 
-        Matcher matcher = PATTERN.matcher(value);
-        if (!matcher.matches()) {
+    static Optional<SlsVersionMatcher> maybeCreate(
+            String value, OptionalInt major, OptionalInt minor, OptionalInt patch) {
+        SlsVersionMatcher maybeMatcher = ImmutableSlsVersionMatcher.of(value, major, minor, patch);
+
+        if (maybeMatcher.getPatchVersionNumber().isPresent()
+                && (!maybeMatcher.getMinorVersionNumber().isPresent()
+                        || !maybeMatcher.getMajorVersionNumber().isPresent())) {
+            // String contains a pattern where major or minor version is underspecified.
+            // Example: x.x.2, 1.x.3, x.2.3
+            log.info(
+                    "Not a valid matcher, a patch version is specified, yet a major or minor is not specified",
+                    SafeArg.of("matcher", maybeMatcher));
+            return Optional.empty();
+        }
+        if (maybeMatcher.getMinorVersionNumber().isPresent()
+                && !maybeMatcher.getMajorVersionNumber().isPresent()) {
+            // String contains a pattern where major version is underspecified. Example: x.2.x
+            log.info(
+                    "Not a valid matcher, a minor version is specified, yet a major is not specified",
+                    SafeArg.of("matcher", maybeMatcher));
```

Comment:
all the logic here is unchanged, it just got pulled out to a `maybeCreate` method so that it's usable from the Regex & new hand-rolled impls.